pkg/s3mapprovider/provider_test.go

```diff
@@ -0,0 +1,303 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3provider
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/confmap"
+	"gopkg.in/yaml.v2"
+)
+
+// Interface testClient standardizes GetObject, which mocks S3 client works
+type testClient struct {
+	GetObject func(context.Context, *s3.GetObjectInput, ...func(*s3.Options)) (*s3.GetObjectOutput, error)
+}
+
+// Create a client mocking S3 client works in normal cases
+func NewTestClient() *testClient {
+	return &testClient{
+		GetObject: func(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error) {
+			// read local config file and return
+			f, err := ioutil.ReadFile("./testdata/otel-config.yaml")
+			if err != nil {
+				return &s3.GetObjectOutput{}, err
+			}
+			return &s3.GetObjectOutput{Body: io.NopCloser(bytes.NewReader(f)), ContentLength: (int64)(len(f))}, nil
+		}}
+}
+
+// testRetrieve: Mock how Retrieve() works in normal cases
+type testRetrieve struct {
+	client *testClient
+}
+
+func NewTestRetrieve() confmap.Provider {
+	return &testRetrieve{client: NewTestClient()}
```

Comment:
```suggestion
func newTestProvider() confmap.Provider {
	return &provider{client: NewTestClient()}
```

Combined with using an interface for `provider.client`, this means that you can eliminate the `testRetrieve` struct and its duplicate `Retrieve()` implementation.  Then you're testing the actual `provider.Retrieve()` implementation without using a real S3 client.  The test `s3Client` implementation can be varied as required for the various test scenarios.

Reply:
Fixed!